File: pilhas-filas/javascript/valid-parentheses.js
function validParentheses(text) {
    let stack = [];
    for (let c of text) {
        if (c === '(') {
            stack.push(')');
        } else if (c === '{') {
            stack.push('}');
        } else if (c === '[') {
            stack.push(']');
        } else if (stack.length === 0 || stack.pop() !== c) {
            return false;
        }
    }
    return stack.length === 0;
}

console.log(validParentheses("()([]{})")); // true
console.log(validParentheses("(){]{}")); // false

File: pilhas-filas/python/valid_parentheses.py
def valid_parentheses(text):
    stack = []
    for c in text:
        if c == '(':
            stack.append(')')
        elif c == '{':
            stack.append('}')
        elif c == '[':
            stack.append(']')
        elif len(stack) == 0 or stack.pop() != c:
            return False
    return len(stack) == 0


print(valid_parentheses("()([]{})"))  # True
print(valid_parentheses("(){]{}"))  # False
